Refactor Optimizer for correctness

stevenyoungs opened the pull request:
When the filter optimizer combines the result of two or more sub-filters, it is currently using the logical_op of the right hand sub-filter. Fix this so that the optimizer uses the parent filter logical_op

Fixes #13799 https://gramps-project.org/bugs/view.php?id=13799
Related forum discussion: https://gramps.discourse.group/t/strange-filter-behavior-in-6-0-1/7505/16

To reproduce using the example Gramps tree
1. create a new filter called "Ancestors of" with a single rule, "Ancestors of <person>"
set ID to I0044, Inclusive ticked
leave the Options set to "all rules must apply"
2. create a new filter called "Siblings of Ancestors" with a single rule, "Siblings of <filter> match"
set the filter to "Ancestors of"
leave the Options set to "all rules must apply"
3. Create a new filter called "The Family and their Spouces" with two rules
a. rule 1: People matching the <filter> and set the filter to "Ancestors of"
b. rule 2: People matching the <filter> and set the filter to "Siblings of Ancestors"
set the Options to "At least one rule must apply"

Filter the people view using the filter "Siblings of Ancestors". 25 people are expected. 0 are shown

The filter "Ancestors of" returns 7 people
The filter "Siblings of Ancestors" returns 18 people
So the union of these two results must, at a minimum, return 18 people. Therefore a result of 0 people cannot be correct.

stevenyoungs commented:
@dsblank I think this is the correct fix but would appreciate your review

dsblank commented:
Fantastic! I'll take a look.

dsblank commented:
Here are the filters in XML:

```xml
<?xml version="1.0" encoding="utf-8"?>
<filters>
  <object type="Person">
    <filter name="Ancestors of" function="and">
      <rule class="IsAncestorOf" use_regex="False" use_case="False">
        <arg value="I0044"/>
        <arg value="1"/>
      </rule>
    </filter>
    <filter name="Family and their Spouses" function="or">
      <rule class="MatchesFilter" use_regex="False" use_case="False">
        <arg value="Ancestors of"/>
      </rule>
      <rule class="MatchesFilter" use_regex="False" use_case="False">
        <arg value="Siblings of Ancestors"/>
      </rule>
    </filter>
    <filter name="Siblings of Ancestors" function="and">
      <rule class="IsSiblingOfFilterMatch" use_regex="False" use_case="False">
        <arg value="Ancestors of"/>
      </rule>
    </filter>
  </object>
</filters>
```

Planning on making this a test in gramps.

Nick-Hall commented:
We should include a unit test with a fix like this.

dsblank commented:
@stevenyoungs I couldn't figure out how to make a PR on your PR. So here is a diff for adding a unittest for the above. It passes after your fix!

```diff
diff --git a/gramps/gen/filters/test/__init__.py b/gramps/gen/filters/test/__init__.py
new file mode 100644
index 000000000..e69de29bb
diff --git a/gramps/gen/filters/test/filter_optimizer_test.py b/gramps/gen/filters/test/filter_optimizer_test.py
new file mode 100644
index 000000000..3fc0df3ca
--- /dev/null
+++ b/gramps/gen/filters/test/filter_optimizer_test.py
@@ -0,0 +1,76 @@
+from gramps.gen.filters import FilterList
+import tempfile
+import os
+import unittest
+
+from ...const import DATA_DIR
+from ...db.utils import import_as_dict
+from ...user import User
+from ...filters import reload_custom_filters
+
+reload_custom_filters()
+
+from ...filters import CustomFilters
+
+
+custom_filters_xml = """<?xml version="1.0" encoding="utf-8"?>
+<filters>
+  <object type="Person">
+    <filter name="Ancestors of" function="and">
+      <rule class="IsAncestorOf" use_regex="False" use_case="False">
+        <arg value="I0044"/>
+        <arg value="1"/>
+      </rule>
+    </filter>
+    <filter name="Family and their Spouses" function="or">
+      <rule class="MatchesFilter" use_regex="False" use_case="False">
+        <arg value="Ancestors of"/>
+      </rule>
+      <rule class="MatchesFilter" use_regex="False" use_case="False">
+        <arg value="Siblings of Ancestors"/>
+      </rule>
+    </filter>
+    <filter name="Siblings of Ancestors" function="and">
+      <rule class="IsSiblingOfFilterMatch" use_regex="False" use_case="False">
+        <arg value="Ancestors of"/>
+      </rule>
+    </filter>
+  </object>
+</filters>
+"""
+TEST_DIR = os.path.abspath(os.path.join(DATA_DIR, "tests"))
+EXAMPLE = os.path.join(TEST_DIR, "example.gramps")
+
+class OptimizerTest(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        """
+        Import example database.
+        """
+        cls.db = import_as_dict(EXAMPLE, User())
+
+        with tempfile.NamedTemporaryFile(mode='w+', encoding='utf8') as tmp_file:
+            tmp_file.write(custom_filters_xml)
+            tmp_file.seek(0)
+            fl = FilterList(tmp_file.name)
+            fl.load()
+
+        the_custom_filters = CustomFilters.get_filters_dict("Person")
+        cls.filters = fl.get_filters_dict("Person")
+        for filter_name in cls.filters:
+            the_custom_filters[filter_name] = cls.filters[filter_name]
+
+    def test_ancestors_of(self):
+        filter = self.filters["Ancestors of"]
+        results = filter.apply(self.db)
+        self.assertEqual(len(results), 7)
+
+    def test_siblings_of_ancestors(self):
+        filter = self.filters["Siblings of Ancestors"]
+        results = filter.apply(self.db)
+        self.assertEqual(len(results), 18)
+
+    def test_family_and_their_spouses(self):
+        filter = self.filters["Family and their Spouses"]
+        results = filter.apply(self.db)
+        self.assertEqual(len(results), 25)
```

Let me know if there is way to do this appropriately.

dsblank reviewed (approved):
Very nice; thanks!

stevenyoungs commented:
hhmmmm I'm not sure why other tests are suddenly failing. I'm glad I made you co-author now @dsblank 🤣

dsblank commented:
I'll take a look :)

dsblank commented:
It is probably because I altered the global CustomFilters. Let me see if I can cleanup after the new tests.

dsblank commented:
Definitely caused by my test. Attempting to fix...

dsblank commented:
Okay, I think I have made the pytest and Gramps gods happy.

Nick-Hall commented:
Please add entries in the `po/POTFILES.skip` file for the new files.

stevenyoungs commented:
> Please add entries in the `po/POTFILES.skip` file for the new files.

Done.

kkujansuu commented:
Sorry to intrude but as far as I can see filtering still does not work in Gramps 6 even with this optimizer fix. See for example the filters below.

The filter F1 obviously should yield all individuals in the database (since it is an "or" filter with the "Everyone" rule).
But it only yields the same results as filter F3:

```
  <object type="Person">
    <filter name="F1" function="or">
      <rule class="Everyone" use_regex="False" use_case="False">
      </rule>
      <rule class="MatchesFilter" use_regex="False" use_case="False">
        <arg value="F2"/>
      </rule>
    </filter>
    <filter name="F2" function="and">
      <rule class="MatchesFilter" use_regex="False" use_case="False">
        <arg value="F3"/>
      </rule>
    </filter>
    <filter name="F3" function="and">
      <rule class="IsAncestorOf" use_regex="False" use_case="False">
        <arg value="I0041"/>
        <arg value="0"/>
      </rule>
    </filter>
  </object>

```
![test](https://github.com/user-attachments/assets/1bc6189a-ae7f-4f95-aa5c-f9b8a28f58a9)

Or have I misunderstood something?

dsblank commented:
This looks like another test to add to the test suite. Thanks!

dsblank commented:
Added @kkujansuu 's additional issue as another set of unittests. Now we have one failure to fix in the optimizer.

stevenyoungs commented:
Here's a tabular version of the nice diagram @kkujansuu created. It includes additional information on the rules which have a `selected_handles` attribute.

|name|rule|logical_op|optimizer|
|--|--|--|--|
|F3|Ancestors of <person> "I0041"|and|selected_handles|
|F2|People matching the <filter> "F3"|and||
|F1|Everyone<br>People matching the <filter> "F2"|or||

The filter optimizer walks the tree of filters, looking for rules where
1. there is a `selected_handles` attribute
2. the logical_op is "and"

Where both conditions are true, the `selected_handles` are used to reduce the search space for any other, non-optimized rules. It reduces the search space by limiting handles which are checked in subsequent rules to those in `selected_handles` (`handles_in`) or for inverted rules, testing all handles in the db, excluding those handles in `selected_handles` (`handles_out`).

Earlier in this PR, I modified test 2 (`logical_op == 'and'`) to check the `logical_op` of the parent filter. The parent filter is F2 in this scenario. As can be seen, this also has a `logical_op` of `'and'`.

From the table above, we can see that the rule in filter F3 meets both conditions (`selected_handles` and `logical_op=='and'` [or `parent_logical_op=='and'`]). The optimizer therefore *only* considers the 16 handles resulting from filter F2 when evaluating the "Everyone" rule. Whilst this is quicker, it is also, unfortunately, wrong.

Alas, I'm still not sure what the correct fix is....

dsblank commented:
@stevenyoungs, I made a change [cc8bc77](https://github.com/gramps-project/gramps/pull/2052/commits/cc8bc775777df1f37a4891ca9d54d68a14a9d968) that I believe fixes the issues, but not 100% sure yet. Need further testing.

dsblank commented:
I don't think the failing tests have anything to do with these changes.

stevenyoungs commented:
> @stevenyoungs, I made a change [cc8bc77](https://github.com/gramps-project/gramps/pull/2052/commits/cc8bc775777df1f37a4891ca9d54d68a14a9d968) that I believe fixes the issues, but not 100% sure yet. 

Like you, I'm not entirely confident, but it feels wrong. `parent_logical_op` is now always going to be the `logical_op` of the root (top-most) filter. i.e. constant for a given execution of Optimizer. So every entry appended to `result` will have the same logical_op. 
`get_handles`, which tests the value of `logical_op` for each value in `result` (now called `all_selected_handles`) will consider everything as "and" or as "or" (because `parent_logical_op` is always the same value). so if one of the lower filters has a logical_op of "or", I believe it would get treated by `get_handles` as "and".

My assumption is that this works for @kkujansuu's example because the root filter is "or". If there was an F0 filter added on top, which used "and", my expectation is that the wrong result will be generated.

>  I think we can make up a tree of AND's and OR's that still fails.

I believe so too. I think we need a multi-level filter, where the filter on each level has multiple rules. We then alternate these filters to use "and" or "or". Then repeat with each filter using "or" and repeat again with each filter using "and". Then for completeness, repeat with negation. @kkujansuu's example leads us in this direction.

dsblank commented:
> Could something like this work?

Yes, I think so. The thing I missed in my first version was the ability to short-circuit any further processing when encountering an "or" at any level.

stevenyoungs commented:
I've taken @kkujansuu's idea and implemented. Initial results look promising.
His example filters from https://github.com/gramps-project/gramps/pull/2052#issuecomment-2834278763 now produce the correct results.

It does not yet handle inverted filters, therefore tests are expected to fail.

stevenyoungs commented:
Implemented support for inverted filters.

I'm not sure that "one" is handled correctly. I think it should use `symmetric_difference()` rather than `union`. `union` will give us results where more than one rule match

stevenyoungs commented:
> I'm not sure that "one" is handled correctly. I think it should use `symmetric_difference()` rather than `union`. `union` will give us results where more than one rule match

whilst the result of Optimizer is not what I expected, the end result is correct due to logic in `GenericFilter `
```
                    if apply_logical_op(db, obj, self.flist) != self.invert:
                        final_list.append(obj.handle)
```

I need to think about that...

stevenyoungs commented:
Something in `OptimizerTest.setUpClass()` and/or `OptimizerTest.tearDownClass()` appears to be causing 9 tests from `gramps.plugins.test.reports_test` to fail.

The errors are all of the form:
```
======================================================================
ERROR: test_WebCal (gramps.plugins.test.reports_test.TestDynamic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/gramps/gramps/gramps/plugins/test/reports_test.py", line 93, in test_method
    self.assertTrue(test_function(out, err, report_name, **options))
  File "/home/runner/work/gramps/gramps/gramps/plugins/test/reports_test.py", line 171, in test_output_file
    raise Exception("can't find '%s' in order to delete it" % filename)
Exception: can't find '/tmp/WEBCAL' in order to delete it
```

I've modified `OptimizerTest` to avoid using a temporary file. I can't see anything else that could be the cause.

stevenyoungs commented:
It appears to be one or both calls to `filters.set_custom_filters` that cause the other tests to fail.
Once the problem is identified the 3 most recent commits can be discarded.

dsblank commented:
I'll take a look at this now.

dsblank commented:
Thank @kkujansuu for the suggestion, and @stevenyoungs for the implementation! I think this is much better not only that it is correct, but because it is easier to understand and debug.

Is this an accurate summary of the changes:

1. Optimizer now starts with the set of all handles, and removes non-matching ones. (This might have a slight impact on memory, but even 100k handles are not that big)
2. Now adds a new optimization for "only one of" types of filters where only one rule should apply
3. Collects the matching handles via "two functions that call each other recursively"

As we move forward, we can probably expand the tests (which all kinds of variations) to prove the correctness.

kkujansuu commented:
I think there are still a couple of special cases that don't work:

1. Empty filter with invert.

```
    <filter name="invert" function="and" invert="1">
    </filter>

```

2. The symmetric difference for "one" filter does not always work, e.g.

```
    <filter name="I0001 or I0002" function="one">
      <rule class="HasIdOf" use_regex="False" use_case="False">
        <arg value="I0001"/>
      </rule>
      <rule class="HasIdOf" use_regex="False" use_case="False">
        <arg value="I0002"/>
      </rule>
    </filter>
```

And I think it is not necessary to generate the "all_handles" sets. Instead one could use a special class instead of the sets. Something like

```
INCLUDE = 1
EXCLUDE = 2

@dataclass
class Handles:
    rectype: int  # INCLUDE or EXCLUDE
    handles: set[str]
    
```
Then 

    Handles(INCLUDE, set1)
    
would represent the set "set1"

    Handles(EXCLUDE, set1)

would represent the inverse of "set1", i.e. all handles not in set1.

    Handles(INCLUDE, set())

would be the empty set and

    Handles(EXCLUDE, set())
    
would represent "all handles".

The functions compute_xxx would return a Handles object instead of a set of handles.

One would also need to define the set operations for the Handles objects, e.g.

so that e.g.

    Handles(INCLUDE, set1) & Handles(INCLUDE, set2) == Handles(INCLUDE, set1 & set2)
    Handles(INCLUDE, set1) & Handles(EXCLUDE, set2) == Handles(INCLUDE, set1 - set2)

Does that make sense?

stevenyoungs commented:
> Thank @kkujansuu for the suggestion, and @stevenyoungs for the implementation! I think this is much better not only that it is correct, but because it is easier to understand and debug.
> 
> Is this an accurate summary of the changes:
> 
> 1. Optimizer now starts with the set of all handles, and removes non-matching ones. (This might have a slight impact on memory, but even 100k handles are not that big)
> 2. Now adds a new optimization for "only one of" types of filters where only one rule should apply
> 3. Collects the matching handles via "two functions that call each other recursively"
> 

Optimizer attempts to reduce the search space of the filter tree. The handles in this, possibly reduced, search space, are then evaluated against the filters (in `GenericFilter`)
Optimizer:
1. recurses the tree of filters in a depth first manner. 
2. When a leaf rule is reached, if returns one of
a. the selected_handles of the rule, if available
or
b. all_handles
3. Once all rules in a filter have been evaluated, (stored in `handlesets`) Optimizer uses the logical_op of the filter to combine the 0, 1 or more sets of results in `handlesets`, using intersection, union or symmetric_difference as appropriate

There some further optimisations that we can make 
1. for "union" if one of the sets in `handlesets` has the same size as `all_handles` we can avoid computing the union and simply return `all_handles`
2. for "interesction" if one of the sets in `handlesets` is empty, we can avoid computing the intersection and return an empty set
We can go further and avoid even calling `compute_potential_handles_for_rule`

> As we move forward, we can probably expand the tests (which all kinds of variations) to prove the correctness.

Yes, I think we can dynamically generate some tests using the example tree and the Everyone and Home Person rules, for example using "and", "or, and "one" rules. Additionally tests with an empty filter and more complex trees of filter using mixes of "and", "or, and "one".

stevenyoungs commented:
> 1. Empty filter with invert.

In Gramps 5.2, a filter with no rules
|logical_op|invert|result|
|---|---|---|
|and|False|all handles|
|and|True|zero handles|
|or|False|zero handles|
|or|True|all handles|
|one|False|zero handles|
|one|True|all handles|

`logical_op` behaves logically, inverting the result in all cases.
It's not entirely obvious why the non-inverted result is as it is. How does a filter with no rules actually match *anything*? Why does changing the `logical_op` vary the result - there is nothing to and\or\one of with.

stevenyoungs commented:
> 2\. The symmetric difference for "one" filter does not always work, e.g.

When a rule does not offer the `selected_handles` optimisation, all_handles is returned.
The "HasIdOf" rules does not offer `selected_handles` (currently) so both rules return all_handles. The symmetric_difference of two equal sets is an empty set. This is obviously wrong. Now we know why, I'll look at a fix.

Nick-Hall commented on gramps/gen/filters/optimizer.py:
```diff
@@ -43,95 +44,73 @@ def union(sets):
         return set()
 
 
+def symmetric_difference(sets):
```
You could write this function as:
```
def symmetric_difference(sets):
    return reduce(lambda x, y: x^y, sets, set())
```

This will need `from functools import reduce` added to the imports.

The same approach could be used for _union_ and _intersection_.

stevenyoungs replied:
Nice suggestion @Nick-Hall!
I've implemented this. For intersection I've avoided computing x&x when we only have a single set.

Nick-Hall commented on gramps/gen/filters/optimizer.py:
```diff
@@ -43,95 +44,73 @@ def union(sets):
+        self.handles_in = self.compute_potential_handles_for_filter(filter)
+        self.handles_out = set()
+
+    def compute_potential_handles_for_filter(self, filter):
```
You should add a docstring for this method.

stevenyoungs replied:
Done

Nick-Hall commented on gramps/gen/filters/optimizer.py:
```diff
@@ -43,95 +44,73 @@ def union(sets):
+
+        return handles
+
+    def compute_potential_handles_for_rule(self, rule):
```
You should add a docstring for this method.

stevenyoungs replied:
Done

Nick-Hall reviewed (changes requested):
Please add the docstrings.  I'll leave the code change decision up to you.

stevenyoungs commented on gramps/gen/filters/_genericfilter.py:
```diff
@@ -144,51 +147,29 @@ def apply_logical_op_to_all(
+            len(handles_in),
         )
-        LOG.debug("Optimizer handles_out: %s", len(handles_out))
         if id_list is None:
```
I think there is a further optimization: when `id_list` is not None, we can precompute the set of handles and then pass this set into Optimizer to use as `all_handles`.
I don't want to pursue this until we have a working implementation.

dsblank replied:
Now that everything is working I can work on this, and the line 29 optimization if any set is empty. (I'll do some speed tests to see if that is worthwhile)

dsblank replied:
Thanks for fixing the id_list speedup, and handling the tuple interface.

stevenyoungs commented on gramps/gen/filters/optimizer.py:
```diff
@@ -18,29 +19,18 @@
-        return result
-    else:
-        return set()
+    return reduce(lambda x, y: x & y, sets, sets[0] if sets else set())
```
if any set in handlesets is empty, we can optimize by immediately returning set()

dsblank replied:
I generalized your basic idea: if you put the shorter sets first, then it will reduce faster.

stevenyoungs replied:
Nice!
Given that you now have a sorted list of sets, for one if statement, you could test if sets[0] has length zero. Whether it is worth it is a different matter!

dsblank replied:
True, but it may be more expensive to do the test :) Anyway, I think this is clean, elegant, and not prone to errors.

stevenyoungs commented:
> 2. The symmetric difference for "one" filter does not always work, e.g.
> 
> ```
>     <filter name="I0001 or I0002" function="one">
>       <rule class="HasIdOf" use_regex="False" use_case="False">
>         <arg value="I0001"/>
>       </rule>
>       <rule class="HasIdOf" use_regex="False" use_case="False">
>         <arg value="I0002"/>
>       </rule>
>     </filter>
> ```

@kkujansuu, this is now working and I've added tests for _one_ (xor), _or_ and _and_ variants of this filter.
Thanks for testing!

stevenyoungs commented:
> 1. Empty filter with invert.
> 
> ```
>     <filter name="invert" function="and" invert="1">
>     </filter>
> ```

Observations of the behaviour in Gramps v5.2 and v6.0 with this PR applied

logical_op | invert | v5.2|v6.0|
-- | -- | --| --|
|and|False|all handles|all handles|
|and|True|zero handles|zero handles|
|or|False|zero handles|zero handles|
|or|True|all handles|all handles|
|one|False|zero handles|zero handles|
|one|True|all handles|all handles|

I've added 6 new unit tests to capture the above.

dsblank commented:
Although the optimization of `intersection` *only* reduced the time from milliseconds to nanoseconds, it was easy and could speedup could accumulate. I tested with large sets of handles, with an empty set at the end. Test code:

```python
import random
from functools import reduce
from gramps.gen.utils.id import create_id

def make_sets(n):
    sets = []
    for i in range(n):
        sets.append(set([create_id() for i in range(random.randint(20, 999_999))]))
    sets.append(set([]))
    return sets

sets = make_sets(10)
```

Without optimization:
```ipython
In [19]: %%time
    ...: functools.reduce(lambda x, y: x & y, sets, sets[0] if sets else set())
    ...: 
    ...: 
CPU times: user 46 ms, sys: 8.99 ms, total: 54.9 ms
Wall time: 54.9 ms
```

With optimization:
```ipython
In [20]: %%time
    ...: functools.reduce(lambda x, y: x & y, sorted(sets, key=lambda s: len(s)), sets[0] if sets else set())
    ...: 
    ...: 
CPU times: user 6 μs, sys: 1 μs, total: 7 μs
Wall time: 8.11 μs
```

The optimization of union could not make such large increases in speed, but I sped it quite a bit (4x) by using a trick to do more processing in C:

Without optimization:
```ipython
In [22]: %%time
    ...: x = reduce(lambda x, y: x | y, sets, set())
    ...: 
    ...: 
CPU times: user 1.79 s, sys: 653 ms, total: 2.45 s
Wall time: 2.45 s
```

With optimization:
```ipython
In [21]: %%time
    ...: x = set.union(*sets)
    ...: 
    ...: 
CPU times: user 289 ms, sys: 144 ms, total: 433 ms
Wall time: 433 ms
```

The optimization when you have an id_list was easy:

```python
        if id_list:
            optimizer = Optimizer(set(id_list), self)
        else:
            optimizer = Optimizer(set(self.get_all_handles(db)), self)
        possible_handles = optimizer.get_possible_handles()
```

stevenyoungs commented on gramps/gen/filters/_genericfilter.py:
```diff
@@ -144,51 +147,34 @@ def apply_logical_op_to_all(
 
-        optimizer = Optimizer(self)
-        handles_in, handles_out = optimizer.get_handles()
+        if id_list:
+            optimizer = Optimizer(set(id_list), self)
+        else:
+            optimizer = Optimizer(set(self.get_all_handles(db)), self)
```
```suggestion
        optimizer = Optimizer(set(id_list if id_list else self.get_all_handles(db)), self) 
```
This might be cleaner

stevenyoungs replied:
There's more to it than this. I'll add a commit

dsblank replied:
Nice catch!

stevenyoungs commented:
sorry @dsblank, I hadn't intended to get in your way!

stevenyoungs commented on gramps/gen/filters/_genericfilter.py:
```diff
@@ -142,76 +145,44 @@ def get_number(self, db):
     def apply_logical_op_to_all(
         self, db, id_list, apply_logical_op, user=None, tupleind=None, tree=False
```
The tree parameter does not appear to be used. Should it be?
It is a parameter of `GenericFilter.apply` from where it is passed straight through

dsblank replied:
I didn't alter the original API. We can add this to a clean up in 6.1 since we are in bug-fix mode on 6.0.

stevenyoungs replied:
in 5.2, `tree` is [used](https://github.com/gramps-project/gramps/blob/0f8d4ecd429431b4df64910962f8764af9ff1766/gramps/gen/filters/_genericfilter.py#L175C13-L175C86)
```
    def check_and(self, db, id_list, user=None, tupleind=None, tree=False):
        final_list = []
        flist = self.flist
        if user:
            user.begin_progress(_("Filter"), _("Applying ..."), self.get_number(db))
        if id_list is None:
            with self.get_tree_cursor(db) if tree else self.get_cursor(db) as cursor:
 ```

and then in `GenericPlaceFilter` this gives different behaviour

```
    def get_cursor(self, db):
        return db.get_place_cursor()

    def get_tree_cursor(self, db):
        return db.get_place_tree_cursor()
 ```
 
 I've changed the behaviour this when I introduced `self.get_all_handles(db)` - although clearly we do not have any tests for this.

stevenyoungs replied:
So far I cannot see any code that passes in the `tree` argument to `apply`. It can only be a filter of which the class or super class is `GenericPlaceFilter`.  `GenericPlaceFilter` is the only type where `get_cursor` and `get_tree_cursor` give different results.

dsblank replied:
I noted the missing code below.

stevenyoungs replied:
Can you find anywhere where the tree parameter is set to True?

stevenyoungs replied:
For the record, the Places view sets tree=True

dsblank commented:
@stevenyoungs I think if you are ready, this can be merged. Ended up with a couple of more optimizations along with the major bug fix. This is worth an immediate 6.0.2

CC: @Nick-Hall 

Nice work everyone!

stevenyoungs commented:
> @stevenyoungs I think if you are ready, this can be merged. Ended up with a couple of more optimizations along with the major bug fix. This is worth an immediate 6.0.2
> 
See my comment above about the `tree` parameter to `apply` and `apply_logical_op_to_all`

The docstring of the return value from `apply` does not match the code. See line 236 of _genericfilter.py
```
        :Returns: if id_list given, it is returned with the items that
                do not match the filter, filtered out.
                if id_list not given, all items in the database that
                match the filter are returned as a list of handles
```

the code always returns the list of handles returned by `apply_logical_op_to_all`.

dsblank commented:
https://github.com/gramps-project/gramps/pull/2052/files#diff-9970918f74dbce48e9687b440d5d9352c5f7909e6f42785ec4ba1ce27f2ff02dL170

dsblank commented:
The only object where get_tree_cursor() is different from get_cursor() is in the `place` filters.

dsblank commented:
And if id_list is used, it needs to return objects (or data?), not handles.

dsblank commented:
I can take a look at what is missing, and add tests.

stevenyoungs commented:
> I can take a look at what is missing, and add tests.

I've added support for the `tree` parameter and in the process discovered that the "Every place" yields a result of nothing :-(

stevenyoungs commented:
1. How on earth did the tests pass when I broke `intersection` when I converted it to use `reduce`?
This reinforces my concern that coverage is not yet good enough
2. I'm concerned that placemodel is relying on the order of results.
So, now that the "Every place" filter returns the correct number of handles, the display is not as expected. I believe that this is down to [`PlaceModel.add_row`](https://github.com/gramps-project/gramps/blob/96f2b6bc04fa248dca85e15763efdb513c9db0ed/gramps/gui/views/treemodels/placemodel.py#L365) expecting parent places to be added to the model _first_.
In the latest version of this PR, place P1354 is the first handle passed to add_row. It does have a parent place. The parent place is not already in the model. So place P1354 is added as a root node. This is incorrect. I would expect one of places [P0441, P0442, P0957] to be the first ones added, these being the true root node countries Greece, Russia, USA. We now store the `possible_handles` in a `set`, which may have a different order than the order in which the handles were retrieved via  `self.get_tree_cursor`.

Nick-Hall commented:
> I'm concerned that placemodel is relying on the order of results.

Yes. That is why we use tree cursors.

stevenyoungs commented:
I see two options to solve this
1. when `tree==True`, do not use `Optimizer` and process handles in the order returned by `get_tree_cursor`
filter performance for Places will be broadly the same as gramps 5.2
2. continue to use Optimizer and then at the end of `apply`, sort the result back into the expected order
filter performance for Places should be better than Gramps 5.2, so long as the gains from `Optimizer` outweigh the final sorting of the result

@dsblank any thoughts?

stevenyoungs commented:
I've coded the second option

> continue to use Optimizer and then at the end of apply_logical_op_to_all, sort the result back into the expected order

The code is shorter than the comment.
Some unit tests of Place filters are required that check not only the expected number of handles are returned but that those handles are also in the expected order.

stevenyoungs commented:
> The docstring of the return value from `apply` does not match the code. See line 236 of _genericfilter.py
> 
> ```
>         :Returns: if id_list given, it is returned with the items that
>                 do not match the filter, filtered out.
>                 if id_list not given, all items in the database that
>                 match the filter are returned as a list of handles
> ```
> 
> the code always returns the list of handles returned by `apply_logical_op_to_all`.

I've added code to return tuples when the input id_list is tuples. We need to check the performance.  A simple test is to go to the Events view and in the search bar set Type=Birth and click Find

> @stevenyoungs I think if you are ready, this can be merged. Ended up with a couple of more optimizations along with the major bug fix. This is worth an immediate 6.0.2

All problems that I'm aware of are now addressed 🤞
This PR has grown considerably, and taken a number of wrong turns. As such it needs careful review and testing.

dsblank commented:
I'm free for the next few hours to look over the changes and add more tests. I can look at the typing conflicts too.

stevenyoungs commented:
> I'm free for the next few hours to look over the changes and add more tests. I can look at the typing conflicts too.

Thanks. I'll try and get the typing fixed but will leave you to review and look at unit tests.

dsblank commented on gramps/gen/filters/_genericfilter.py:
```diff
@@ -300,6 +324,9 @@ class GenericFamilyFilter(GenericFilter):
     def __init__(self, source=None):
         GenericFilter.__init__(self, source)
 
+    def get_all_handles(self, db):
```
You can actually avoid having to add another method to all of the filter types.

dsblank commented on gramps/gen/filters/_genericfilter.py:
```diff
@@ -282,9 +278,37 @@ def apply(self, db, id_list=None, tupleind=None, user=None, tree=False):
+            tree_handles = [handle for handle, obj in self.get_tree_cursor(db)]
+            possible_handles = set(tree_handles)
+        else:
+            possible_handles = set(self.get_all_handles(db))
```
We already have a method specific to the filter type. Let's use it, and make it parallel with the tree case:

```suggestion
            all_handles = [handle for handle, obj in self.get_cursor(db)]
            possible_handles = set(all_handles)
```

stevenyoungs replied:
Speed was my concern. `get_all_handles` is O(1) which is better the O(n) of using a `cursor`. The `cursor` also constructs the object which we do not need at this stage.

stevenyoungs replied:
Unless you feel strongly, I will leave the new `get_all_handles` methods

dsblank replied:
Yes, you have a point. Actually I think they are both O(n), but one is in SQL and the other is in Python. Let's leave it as you have it.

dsblank reviewed (changes requested):
You can remove all of the get_all_handles() methods.

stevenyoungs commented:
`IsDescendantFamilyOf` was adding `None` to it's list of `selected_handles`.  Preventing this allows GenericFilter to also avoid remove `None` from `possible_handles`. Conceptually this is the right change. It remains to be seen if other files add `None` to their `selected_handles`

dsblank commented:
This is a clever and efficient method to not lose the order of the id_list, considering it can only shrink (be filtered) in the result.

```python
res = sorted(res, key=lambda x: tree_handles.index(x))
```

dsblank commented:
A possibility occurs to me (for Gramps 6.1):  if all of the filter.prepare() methods were lazy (eg used set comprehension rather than sets) then we needn't compute the actual members until needed. But that would remove the optimization based on set length (the sort). But could be worth it overall.

stevenyoungs commented:
> This is a clever and efficient method to not lose the order of the id_list, considering it can only shrink (be filtered) in the result.
> 
> ```python
> res = sorted(res, key=lambda x: tree_handles.index(x))
> ```

My only doubt is performance, but we can optimise by building a Dict[PrimaryObjectHandle, index] if so.

stevenyoungs commented:
> A possibility occurs to me (for Gramps 6.1): if all of the filter.prepare() methods were lazy (eg used set comprehension rather than sets) then we needn't compute the actual members until needed. But that would remove the optimization based on set length (the sort). But could be worth it overall.

And some other ideas:
1. For some rules, rule.selected_handles is the exact result. Currently we ignore that and still call rule.apply_to_one in GenericFilter
2. prepare currently works across the entire db. If it knew the starting possible_handles, it may be able to reduce the scope of it's work.

Let's get 6.0 working first!

Nick-Hall commented:
> Let's get 6.0 working first!

Yes.  You should keep the scope of this PR to the bug fixes.  I'm not too worried about performance.

> This PR has grown considerably, and taken a number of wrong turns. As such it needs careful review and testing.

I agree.  This is a much bigger change than we would normally accept in a maintenance branch.

We need to test every rule.  Are there any rules without unit tests?

Then we need to test combinations of rules in filters.

After that we need to do more testing, and then even more testing.

dsblank commented:
I just added tests for id_list, tupleind, and tree.

dsblank commented:
> We need to test every rule. Are there any rules without unit tests?

Every rule has a test.

> After that we need to do more testing, and then even more testing.

I believe that we have tests for all of the options. Of course we can continue to add tests (there are a large combination of rules x options). We also now have tests for things that have NEVER been tested before. 

But we also have to balance that fact with the issue that we currently have a broken release.

I would suggest that we test and us this for a couple of more days and get a release out sooner than later. This is a serious breakage.

Nick-Hall commented on gramps/gen/filters/test/filter_optimizer_test.py:
```diff
@@ -183,9 +182,12 @@ def test_everyone_with_id_list_tupleind_3(self):
         filter = self.filters["Everyone"]
         default_person = self.db.get_default_person()
         id_list = [(default_person, default_person.handle)]
-        with pytest.raises(IndexError) as excinfo:
```
You can use [self.assertRaises](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertRaises) here.

dsblank reviewed (approved):
LGTM!

dsblank commented on gramps/gen/filters/optimizer.py:
```diff
@@ -18,120 +19,86 @@
+    # with interrsection, the shortest of the starting sets determines the
+    # maximum size of the result
+    sorted_sets = sorted(sets, key=lambda s: len(s))
+    return reduce(lambda x, y: x & y, sorted_sets[1:], sorted_sets[0])
```
This is equally fast and perhaps a little easier to read:

```suggestion
    return set.intersection(*sorted_sets)
```

And then we also don't have to import/use functools.

stevenyoungs commented:
Another bug: https://gramps-project.org/bugs/view.php?id=13799#c68867

This is calling apply with id_list, tupleind == 1, tree == False
I'm guessing that the problem is the order of results in the scenario.

stevenyoungs commented on gramps/gen/filters/_genericfilter.py:
```diff
@@ -282,9 +278,37 @@ def apply(self, db, id_list=None, tupleind=None, user=None, tree=False):
+
+        # build the starting set of possible_handles to be filtered
+        possible_handles: Set[PrimaryObjectHandle]
+        if id_list is not None:
+            if tupleind is not None:
```
```suggestion
        if id_list:
            if tupleind is not None:
```
Good catch on tupleind to avoid problems when it is zero. I'm not sure we need to use `is not None` for id_list though

stevenyoungs commented:
> Another bug: https://gramps-project.org/bugs/view.php?id=13799#c68867
> 
> This is calling apply with id_list, tupleind == 1, tree == False I'm guessing that the problem is the order of results is different to the input order in this scenario.

This is now working correctly.

stevenyoungs commented:
Hopefully we now have the holy grail of performance and correctness.
I've no further changes planned and this is ready for review from my perspective.

dsblank commented:
I won't have further time to add more tests for awhile, and this looks very solid to me.

stevenyoungs commented:
> Found another anomaly.
> 
> On the example database, LAGvZ is the only person with the tag "ToDo".
> 
> Build the custom filter: People with the Tag="ToDo" and set the options to **Return values that do not match the filter rules**
> 
> Running the filter returns zero (0) matches.

I can reproduce this.
With **Return values that do not match the filter rules** _unticked_ the correct result of 1 person (LAGvZ) is returned.
With **Return values that do not match the filter rules** _ticked_, no people are returned.

The error originates in Optimizer method `compute_potential_handles_for_filter`
```        
        if filter.invert:
            handles = self.all_handles.difference(handles)
```

`handles` and `self.all_handles` are the same so the result of calling `difference` is empty.

Working backwards, `compute_potential_handles_for_rule` returns `self.all_handles`, because this rule is not currently optimized.
I'm not sure it is safe to carry out the "invert" calculation where we currently do. There are two scenarios where it will give the wrong result
1. a rule is not optimized
this is the current scenario. we end up discarding all handles
2. a rule is optimized but the `selected_handles` is an over estimate
we will discard those handles from `selected_handles` which do not match the rule(s) [and are discarded early because of the invert]

In both cases, we will end up discarding handles too early

dsblank commented:
@stevenyoungs do you want me to add the tests that fail, or do you have them coded already?

stevenyoungs commented:
> @stevenyoungs do you want me to add the tests that fail, or do you have them coded already?

That would be great. I've not done any more than the investigation in my comment above so far.

stevenyoungs commented:
> > @stevenyoungs do you want me to add the tests that fail, or do you have them coded already?
> 
> That would be great. I've not done any more than the investigation in my comment above so far.

@dsblank I've added a couple of additional tests, one of which should fail at the moment
I then have a potential fix to push

stevenyoungs commented:
> Found another anomaly.
> 
> On the example database, LAGvZ is the only person with the tag "ToDo".
> 
> Build the custom filter: People with the Tag="ToDo" and set the options to **Return values that do not match the filter rules**
> 
> Running the filter returns zero (0) matches.

Good news first
* the above problem is now fixed.

Now the bad news
* I'm not particularly happy with the fix. 
Something feels wrong and then....
* the code is still vulnerable if a rule is optimised and provides a superset of the potentially matching handles.
For example if I optimize the Tag rule we used in this problem such that it always sets `selected_handles` to every handle in the DB. This matches the specification; `selected_handles` is a superset of the actual rule result. But when we invert inside optimizer, we will discard all handles and essentially recreate @DaveSch-gramps problem.
* If we keep the contract that a rules `selected_handles` is a superset of the handles matching the rule, then I'm not sure that we can get invert working correctly within Optimizer.

dsblank commented:
I have time today to look at this

DaveSch-gramps commented:
The last fix works the main rule is the filter to return results that are NOT the rule.

Create a filter to find positive results. Then Create a second filter that takes the first filter to find the negative results. No results are returned for the second filter.

These custom filters were created and tested on the example database.

<filters>
  <object type="Person">
    <filter name="Family Name" function="and">
      <rule class="HasNameOf" use_regex="False" use_case="False">
        <arg value=""/>
        <arg value=""/>
        <arg value=""/>
        <arg value=""/>
        <arg value=""/>
        <arg value=""/>
        <arg value=""/>
        <arg value="Garner"/>
        <arg value=""/>
        <arg value=""/>
        <arg value=""/>
      </rule>
    </filter>
    <filter name="NOT Family Name" function="and" invert="1">
      <rule class="MatchesFilter" use_regex="False" use_case="False">
        <arg value="Family Name"/>
      </rule>
    </filter>
  </object>
</filters>

DaveSch-gramps commented:
I had a hard time triaging the previous post. I have a large database (239K people) and running even some of the simplest filters were taking 30+ minutes to resolve/display the list of people. Gramps was displaying "Not Responding". Actually running the filter was not the issue. Running the same filters on 5.2.4 results were returned 2-3 seconds after the final final "applying filter" progress bar was completed.

dsblank commented:
> I had a hard time triaging the previous post. I have a large database (239K people) and running even some of the simplest filters were taking 30+ minutes to resolve/display the list of people. Gramps was displaying "Not Responding". Actually running the filter was not the issue. Running the same filters on 5.2.4 results were returned 2-3 seconds after the final "applying filter" progress bar was completed.

Thanks @DaveSch-gramps for the info. The optimizer should never in any case take longer, so there must be a bug. I'm adding your tests and that will help track the issue down.

dsblank commented:
I think this version properly uses the new recursive calls. I'll have to think about what we can do with "one" and "or". I tested also to make sure that all of the items that were previously optimized still are.

emyoulation commented:
As one of the "others", let me express my gratitude.

stevenyoungs commented on gramps/gen/filters/optimizer.py:
```diff
@@ -17,121 +18,93 @@
+    def compute_potential_handles_for_filter(
+        self, filter: GenericFilter
+    ) -> Tuple[
+        Union[Set[PrimaryObjectHandle], None], Union[Set[PrimaryObjectHandle], None]
```
```suggestion
        Set[PrimaryObjectHandle] | None, Set[PrimaryObjectHandle] | None
```

In other places, `|` is used in preference to the more verbose `Union[]`

stevenyoungs commented on gramps/gen/filters/optimizer.py:
```diff
@@ -17,121 +18,93 @@
+        self,
+        rule: Rule,
+    ) -> Tuple[
+        Union[Set[PrimaryObjectHandle], None], Union[Set[PrimaryObjectHandle], None]
```
```suggestion
        Set[PrimaryObjectHandle] |None, Set[PrimaryObjectHandle] | None
```

stevenyoungs commented on gramps/gen/filters/optimizer.py:
```diff
@@ -17,121 +18,93 @@
+# Typing modules
+#
+# -------------------------------------------------------------------------
+from typing import List, Set, TYPE_CHECKING, Tuple, Union
```
```suggestion
from typing import List, Set, TYPE_CHECKING, Tuple
```

stevenyoungs commented on gramps/gen/filters/optimizer.py:
```diff
@@ -54,79 +54,53 @@ def union(sets: List[Set[PrimaryObjectHandle]]) -> Set[PrimaryObjectHandle]:
+                    if handles_out is None:
+                        handles_out = rule_out
+                    else:
+                        handles_out.intersection(rule_out)
```
Should this be union?
Given a filter with two rules and logical_op "and"
rule 1 says handle 'a' is out. So the path `handles_out = rule_out` is executed
rule 2 says handle 'b' is out. So the path `handles_out.intersection(rule_out)` is executed. This expands to 'a'.intersection('b') which is obviously an empty set

Logically, neither of handles 'a' or 'b' can be in the result so I would expect `handles_out` for the filter to be the set ('a', 'b')

stevenyoungs replied:
I think it will produce the correct result as it is, but `GenericFilter` will call `apply_logical_op` for handles 'a' and 'b' when it doesn't really need to

dsblank replied:
> Should this be union?

I think intersection is correct, because the set can only get smaller.

dsblank replied:
> I think it will produce the correct result as it is, but GenericFilter will call apply_logical_op for handles 'a' and 'b' when it doesn't really need to

Not sure I follow this. You want to make a suggestion?

dsblank replied:
> This expands to 'a'.intersection('b') which is obviously an empty set

I think it is the same set. I think intersection is correct.

```ipython
In [1]: a = {1, 2, 3}
In [2]: b = {1, 2, 3}
In [3]: a.intersection(b)
Out[3]: {1, 2, 3}
```

stevenyoungs commented:
> Thanks @DaveSch-gramps! With the last fix, and more tests, I think we finally have all cases covered.

Brave words given the history of this bug and PR - but I'm also feeling optimistic now.

stevenyoungs commented:
`test_not_i0001_and_i0002` uses the filter "not (I0001 and I0002)" which in turn requires two sub-filters "I0001" and "I0002"
These sub-filters did not exist, yet the unit tests were still passing. Messages were generated in the build output
```
Can't find filter I0001 in the defined custom filters
Can't find filter I0002 in the defined custom filters
```

I've fixed the immediate problem by defining the filters "I0001" and "I0002". In my view the unit test should also have failed.

dsblank commented:
> I've fixed the immediate problem by defining the filters "I0001" and "I0002". In my view the unit test should also have failed.

Good catch. Looking into how the tests could have passed...

dsblank commented:
> Looking into how the tests could have passed...

I see... it returns all handles, which is correct. I'll add more tests where "all" doesn't necessarily give a good answer.

stevenyoungs commented:
> @stevenyoungs if you test with either #2071 or #2062 you'll see that 12 tests fail. I like the idea, but we should probably wait for 6.1 to add features to this PR.

Good find. I've added a test that catches this scenario. The build has just failed, due to this test, so it is working.
Now I'll reorder the commits so we have the test first.
